crates/dbsp/src/circuit/dbsp_handle.rs

```diff
@@ -15,18 +21,159 @@ use std::{
 #[cfg(doc)]
 use crate::circuit::circuit_builder::Stream;
 
+/// A host for some workers in the [`Layout`] for a multi-host DBSP circuit.
+#[derive(Clone, Decode, Encode, Serialize, Deserialize)]
+pub struct Host {
+    /// The IP address and TCP port on which the host listens and to which the
+    /// other hosts connect.
+    pub address: SocketAddr,
+
+    /// The worker thread IDs implemented on this host.  Worker thread IDs start
+    /// with 0 in the first host and increase sequentially from there.  A host
+    /// has `workers.len()` workers.
+    pub workers: Range<usize>,
+
+    _private: (),
+}
+
+impl Debug for Host {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Host")
+            .field("address", &self.address)
+            .field("workers", &self.workers)
+            .finish()
+    }
+}
+
+/// How a DBSP circuit is laid out across one or more machines.
+#[derive(Clone, Debug, Decode, Encode, Serialize, Deserialize)]
+pub enum Layout {
+    /// A layout whose workers run on a single host.
+    Solo {
```

Comment:
this is asymmetric. I would expect either the multi to contain a solo, or Multi with 1 entries to implement correctly solo. This suggests that you are optimizing for the single-host case. This information is already implicit in workers, isn't it?

Reply:
"Unless you mean, keep a local SocketAddr in addition to the Vec<Host>, instead of a local index? But I don't see any advantage to that."

Yeah this is what I was asking about. :) It makes things slightly easier when dealing with cluster managers down the line, in that it's easier to automate supplying each pod a static `Vec<Host>` + local `SocketAddr`. Not a deal breaker though, I could add a wrapper script that derives the indexes before starting the pipeline processes.

"Could we use (DNS name, port) instead of (IP, port)/SocketAddr? IPs won't be stable in a cluster manager setting."

Thoughts on this? It would require replacing `SocketAddr` with a different type, starting with how a Layout is initialized.

` pub fn new_multihost(params: Vec<(SocketAddr, usize)>, local_host_idx: usize) -> Layout {`

Reply:
> > Unless you mean, keep a local SocketAddr in addition to the Vec, instead of a local index? But I don't see any advantage to that.
> 
> Yeah this is what I was asking about. :) It makes things slightly easier when dealing with cluster managers down the line, in that it's easier to automate supplying each pod a static `Vec<Host>` + local `SocketAddr`. Not a deal breaker though, I could add a wrapper script that derives the indexes before starting the pipeline processes.

I don't see why one would change the struct for this reason. It would be a waste to repeatedly search the vector of hosts for one's own SocketAddr.

If you want to supply the host's SocketAddr instead of the index, then I guess we can change the constructor like this:

```rust
    pub fn new_multihost(params: Vec<(SocketAddr, usize)>, local_address: SocketAddr) -> Layout {
        let local_host_idx = params
            .iter()
            .position(|(address, _)| *address == local_address)
            .unwrap();
```

> > Could we use (DNS name, port) instead of (IP, port)? IPs won't be stable in a cluster manager setting.
> 
> Thoughts on this? It would require replacing `SocketAddr` with a different type, starting with how a Layout is initialized.
> 
> ` pub fn new_multihost(params: Vec<(SocketAddr, usize)>, local_host_idx: usize) -> Layout {`

I'm not yet trying to build an API for this case yet. This API only handles a static configuration. The caller can do the name lookup.

In a next step where we want to handle scale-up or scale-down, we'll need an API for changing the layout (which is a term I'm not tied to or anything, it's just the first one that came to mind). Something will need to figure out when changes are needed and then notify all of the hosts about the change. When it notifies them, it might as well give them the IP addresses of who it should talk to; I don't know an advantage to making each host do the DNS lookups individually (what if they got different answers?).

Reply:
If you read the Google map-reduce paper you will notice that they had to hack the DNS server to do batch lookups because it became the bottleneck in launching jobs. Imagine all the clients at once issuing DNS requests for all other clients! In Dryad the controller directly distributed IP addresses to the clients.

Reply:
> I'm not yet trying to build an API for this case yet. This API only handles a static configuration. The caller can do the name lookup.

Ack! We can make the caller handle this and convert them to SocketAddr.

Just so you know where I'm coming from this on this: cluster managers like Kubernetes make it annoying for pods to refer to each other by IP addresses. What is stable when we create a pod through the usual Kubernetes abstractions are their DNS names. The DNS names can be mechanically derived from the StatefulSet and pod names, which we'll know before the pods are launched, whereas IP addresses are only assigned after the pod is created. That makes it easy to create the pipeline container launch commands statically when using DNS names, but not when using IP addresses. Hence the simplification from using DNS names to help workers find each other.

> If you read the Google map-reduce paper you will notice that they had to hack the DNS server to do batch lookups because it became the bottleneck in launching jobs. Imagine all the clients at once issuing DNS requests for all other clients! In Dryad the controller directly distributed IP addresses to the clients.

We are very unlikely to have as many pipeline workers as tasks in a map-reduce setting. I really doubt each pod caching a few DNS look ups once in a while will be a bottleneck.

> In a next step where we want to handle scale-up or scale-down, we'll need an API for changing the layout (which is a term I'm not tied to or anything, it's just the first one that came to mind). Something will need to figure out when changes are needed and then notify all of the hosts about the change. When it notifies them, it might as well give them the IP addresses of who it should talk to; I don't know an advantage to making each host do the DNS lookups individually (what if they got different answers?).

Yes we'll need to design this next step carefully. "Layout" is perfectly fine as a term btw (CorfuDB called it that too). 

The general problem is one of membership management and reconfiguring a distributed system. 

There are a few things to get right here when we get to that step:
* Externally triggered layout changes (scale up/down). Kubernetes can help with this.
* Failure detection. What should a worker do when it can't reach a peer in the middle of a computation?
* Will we have a coordinator process? E.g., one convention is to make the first process be the coordinator as well. Rapid for example.

One thing that we'll need then is to keep network identities separate from process identities. Typically, this is done in distributed systems by having each process have a process ID and an "incarnation ID". So, say worker A is talking to worker B. B has incarnation ID X. Now let's say B restarted, in which case it comes back on incarnation ID X'. Workers always keep an eye on what incarnation of a peer they are talking to (when sending and/or receiving messages). This way, A can detect when it is talking to the restarted instance of B. Usually, systems trigger some kind of recovery/catch-up protocol when they detect new incarnations (e.g., A and B could check that B didn't miss out on any new inputs maybe).

Reply:
I pushed a change that changes the constructor the way you suggested.